lisa/sut_orchestrator/hyperv/source.py

```diff
@@ -0,0 +1,190 @@
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return SourceSchema
+
+    def download(self, server: RemoteNode) -> List[PureWindowsPath]:
+        raise NotImplementedError()
+
+
+class LocalSource(Source):
+    def __init__(self, runbook: LocalSourceSchema) -> None:
+        super().__init__(runbook)
+        self.local_runbook: LocalSourceSchema = self.runbook
+        self._log = get_logger("local", self.__class__.__name__)
+        self.downloaded_files: Optional[List[PureWindowsPath]] = None
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "local"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return LocalSourceSchema
+
+    def download(self, server: RemoteNode) -> List[PureWindowsPath]:
+        if not self.downloaded_files:
+            self.downloaded_files = []
+
+        for file in self.local_runbook.files:
+            self.downloaded_files += self._download_file(file, server)
+
+        return self.downloaded_files
+
+    def _download_file(
+        self, file: SourceFileSchema, server: RemoteNode
+    ) -> List[PureWindowsPath]:
+        local_path = Path(file.source)
+
+        local_files = []
+        if local_path.is_dir():
+            local_files = [f for f in local_path.iterdir() if f.is_file()]
+        else:
+            local_files = [local_path]
+
+        downloaded_paths: List[PureWindowsPath] = []
+
+        for f in local_files:
+            downloaded_paths += self._download_single_file(
+                f, file.destination, file.unzip, server
+            )
+
+        return downloaded_paths
+
+    def _download_single_file(
+        self,
+        local_path: Path,
+        destination: Optional[str],
+        unzip: bool,
+        server: RemoteNode,
+    ) -> List[PureWindowsPath]:
+        if unzip:
+            return self._download_zipped_file(local_path, destination, server)
+
+        destination_is_dir = destination and (
+            destination.endswith("/") or destination.endswith("\\")
+        )
+
+        defined_destination = None
+        if destination:
+            if PureWindowsPath(destination).is_absolute():
+                defined_destination = PureWindowsPath(destination)
+            else:
+                defined_destination = PureWindowsPath(server.working_path / destination)
+
+        dest_path = PureWindowsPath(
+            defined_destination
+            if defined_destination
+            else server.working_path / "sources" / local_path.name
+        )
+
+        if destination_is_dir:
+            dest_path = dest_path / local_path.name
+
+        server.tools[Mkdir].create_directory(
+            str(dest_path) if destination_is_dir else str(dest_path.parent)
+        )
+
+        self._log.debug(f"Copying {local_path} to server")
+        server.shell.copy(local_path, dest_path)
+        self._log.debug("Finished copying.")
+
+        return [dest_path]
```

Comment:
Because this method is to return "single file", so don't return a list.

Reply:
That single file can be a zip file which may have multiple files inside.